File: 의진/프로그래머스/광고삽입.py
def time_to_sec(time):
    h, m, s = map(int, time.split(":"))
    return 60 * 60 * h + 60 * m + s


def sec_to_time(sec):
    h = sec // 3600
    m = (sec % 3600) // 60
    s = (sec % 3600) % 60
    return f"{str(h).zfill(2)}:{str(m).zfill(2)}:{str(s).zfill(2)}"


def solution(play_time, adv_time, logs):
    pt = time_to_sec(play_time)
    diff = [0] * (pt + 2)
    time_block = [0] * (pt + 1)
    accu_time_block = [0] * (pt + 1)
    max_val = 0
    max_i = -1

    for log in logs:
        log = log.split("-")
        diff[time_to_sec(log[0])] += 1
        diff[time_to_sec(log[1])] -= 1

    for i in range(len(time_block)):
        if i == 0:
            time_block[i] = diff[i]
        else:
            time_block[i] = time_block[i - 1] + diff[i]

    for i in range(len(accu_time_block) - 1):
        accu_time_block[i + 1] = accu_time_block[i] + time_block[i]

    # try:
    #     print(accu_time_block[1549:1552])
    # except:
    #     pass

    for i in range(pt):
        if i + time_to_sec(adv_time) >= len(accu_time_block):
            break
        value = accu_time_block[i + time_to_sec(adv_time)] - accu_time_block[i]
        if max_val < value:
            # print(sec_to_time(i))
            max_val = value
            max_i = i

    answer = sec_to_time(max_i)
    return answer

File: 명준/광고삽입.kt
//
//
//fun main() = with(System.`in`.bufferedReader()) {
//
//    val play_time = "02:03:55"
//    val adv_time = "00:14:15"
//    val logs : Array<String> = arrayOf(
//        "01:20:15-01:45:14",
//        "00:40:31-01:00:00",
//        "00:25:50-00:48:29",
//        "01:30:59-01:53:29",
//        "01:37:44-02:02:30"
//    )
//
//    var answer: String = ""
//    var maxViewers = 0
//    var bestStart = 0
//
//    val playTimeSec = timeToSec(play_time)
//    val adTimeSec = timeToSec(adv_time)
//
//    val viewers = IntArray(playTimeSec + 1)
//
//    for (i in logs) {
//        val (s,e) = i.split("-").map { timeToSec(it) }
//
//        for (j in s until e) {
//            viewers[j]++
//        }
//    }
//
//    for (j in 0.. playTimeSec - adTimeSec) {
//        val currentViewers = viewers.sliceArray(j until j + adTimeSec).sum()
//        if (currentViewers > maxViewers) {
//            maxViewers = currentViewers
//            bestStart = j
//        }
//    }
//
//    answer = secToTime(bestStart)
//
//}
//
//fun timeToSec(time:String) : Int {
//    val (h, m, s) = time.split(":").map { it.toInt() }
//    return h * 3600 + m * 60 + s
//}
//
//fun secToTime(time:Int) : String {
//    val h = time / 3600
//    val m = (time % 3600) / 60
//    val s = time % 60
//    return "%02d:%02d:%02d".format(h, m, s)
//}